src/row_metadata.rs

```diff
@@ -0,0 +1,195 @@
+// FIXME: this is allowed only temporarily,
+//  before the insert RBWNAT implementation is ready,
+//  cause otherwise the caches are never used.
+#![allow(dead_code)]
+#![allow(unreachable_pub)]
+
+use crate::row::RowKind;
+use crate::sql::Identifier;
+use crate::Result;
+use crate::Row;
+use clickhouse_types::{parse_rbwnat_columns_header, Column};
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::sync::Arc;
+use tokio::sync::{OnceCell, RwLock};
+
+/// Cache for [`RowMetadata`] to avoid allocating it for the same struct more than once
+/// during the application lifecycle. Key: fully qualified table name (e.g. `database.table`).
+type LockedRowMetadataCache = RwLock<HashMap<String, Arc<RowMetadata>>>;
+static ROW_METADATA_CACHE: OnceCell<LockedRowMetadataCache> = OnceCell::const_new();
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum AccessType {
+    WithSeqAccess,
+    WithMapAccess(Vec<usize>),
+}
+
+/// [`RowMetadata`] should be owned outside the (de)serializer,
+/// as it is calculated only once per struct. It does not have lifetimes,
+/// so it does not introduce a breaking change to [`crate::cursors::RowCursor`].
+pub(crate) struct RowMetadata {
+    /// Database schema, or columns, are parsed before the first call to (de)serializer.
+    pub(crate) columns: Vec<Column>,
+    /// This determines whether we can just use [`crate::rowbinary::de::RowBinarySeqAccess`]
+    /// or a more sophisticated approach with [`crate::rowbinary::de::RowBinaryStructAsMapAccess`]
+    /// to support structs defined with different fields order than in the schema.
+    /// (De)serializing a struct as a map will be approximately 40% slower than as a sequence.
+    access_type: AccessType,
+}
+
+impl RowMetadata {
+    pub(crate) fn new<T: Row>(columns: Vec<Column>) -> Self {
+        let access_type = match T::KIND {
+            RowKind::Primitive => {
+                if columns.len() != 1 {
+                    panic!(
+                        "While processing a primitive row: \
+                        expected only 1 column in the database schema, \
+                        but got {} instead.\n#### All schema columns:\n{}",
+                        columns.len(),
+                        join_panic_schema_hint(&columns),
+                    );
+                }
+                AccessType::WithSeqAccess // ignored
+            }
+            RowKind::Tuple => {
+                if T::COLUMN_COUNT != columns.len() {
+                    panic!(
+                        "While processing a tuple row: database schema has {} columns, \
+                        but the tuple definition has {} fields in total.\
+                        \n#### All schema columns:\n{}",
+                        columns.len(),
+                        T::COLUMN_COUNT,
+                        join_panic_schema_hint(&columns),
+                    );
+                }
+                AccessType::WithSeqAccess // ignored
```

Comment:
In the case of `(Row, T)` it's impossible to get `WithMapAccess`, can be considered a bug

Reply:
will address in a follow-up